File: clients/ios/BuildIt/Modules/Calling/ChannelEscalation.swift
// ChannelEscalation.swift
// BuildIt - Decentralized Mesh Communication
//
// Handles transitions between messaging and voice channels.
// Maintains thread/call linkage for seamless channel switching.

import Foundation
import Combine
import os.log

/// Direction of channel escalation
public enum EscalationDirection: String, Codable, Sendable {
    case toVoice = "to-voice"
    case toMessaging = "to-messaging"
}

/// Who initiated the escalation
public enum EscalationInitiator: String, Codable, Sendable {
    case `operator`
    case caller
    case system
}

/// Status of an escalation request
public enum EscalationStatus: String, Codable, Sendable {
    case pending
    case accepted
    case declined
    case completed
    case failed
}

/// Escalation request for channel switching
public struct EscalationRequest: Codable, Sendable, Identifiable {
    public let id: String
    public var threadId: String
    public var callId: String?
    public let direction: EscalationDirection
    public let initiatedBy: EscalationInitiator
    public var reason: String?
    public var status: EscalationStatus
    public let createdAt: Int
    public var completedAt: Int?

    public init(
        id: String = UUID().uuidString,
        threadId: String,
        callId: String? = nil,
        direction: EscalationDirection,
        initiatedBy: EscalationInitiator,
        reason: String? = nil,
        status: EscalationStatus = .pending,
        createdAt: Int = Int(Date().timeIntervalSince1970),
        completedAt: Int? = nil
    ) {
        self.id = id
        self.threadId = threadId
        self.callId = callId
        self.direction = direction
        self.initiatedBy = initiatedBy
        self.reason = reason
        self.status = status
        self.createdAt = createdAt
        self.completedAt = completedAt
    }
}

/// Result of an escalation operation
public struct EscalationResult: Sendable {
    public let success: Bool
    public let escalationId: String
    public var newCallId: String?
    public var newThreadId: String?
    public var error: String?

    public init(
        success: Bool,
        escalationId: String,
        newCallId: String? = nil,
        newThreadId: String? = nil,
        error: String? = nil
    ) {
        self.success = success
        self.escalationId = escalationId
        self.newCallId = newCallId
        self.newThreadId = newThreadId
        self.error = error
    }
}

/// Escalation event for observation
public enum EscalationEvent: Sendable {
    case started(EscalationRequest)
    case requested(request: EscalationRequest, callerPubkey: String)
    case accepted(EscalationRequest)
    case declined(EscalationRequest)
    case completed(EscalationRequest, callId: String?, threadId: String?)
    case failed(EscalationRequest, String)
    case linkCleared(channelId: String)
}

/// Service for handling channel escalation between messaging and voice
@MainActor
public class ChannelEscalation: ObservableObject {
    // MARK: - Published Properties

    @Published public private(set) var escalations: [String: EscalationRequest] = [:]
    @Published public private(set) var pendingEscalations: [EscalationRequest] = []

    // MARK: - Properties

    private var threadToCall: [String: String] = [:] // threadId -> callId
    private var callToThread: [String: String] = [:] // callId -> threadId
    private var operatorPubkey: String = ""
    private let logger = Logger(subsystem: "com.buildit", category: "ChannelEscalation")

    /// Event stream for escalation changes
    public let eventSubject = PassthroughSubject<EscalationEvent, Never>()

    // Weak references to related managers
    private weak var messagingQueueManager: MessagingQueueManager?
    private weak var callingService: CallingService?

    // MARK: - Initialization

    public init() {}

    /// Initialize with operator context and related services
    public func initialize(
        operatorPubkey: String,
        messagingQueueManager: MessagingQueueManager? = nil,
        callingService: CallingService? = nil
    ) {
        self.operatorPubkey = operatorPubkey
        self.messagingQueueManager = messagingQueueManager
        self.callingService = callingService
        logger.info("Initialized ChannelEscalation")
    }

    // MARK: - Escalate to Voice

    /// Escalate a messaging thread to a voice call
    public func escalateToVoice(
        threadId: String,
        thread: MessagingThread? = nil,
        reason: String? = nil
    ) async -> EscalationResult {
        let escalationId = UUID().uuidString
        let now = Int(Date().timeIntervalSince1970)

        var escalation = EscalationRequest(
            id: escalationId,
            threadId: threadId,
            direction: .toVoice,
            initiatedBy: .operator,
            reason: reason,
            status: .pending,
            createdAt: now
        )

        escalations[escalationId] = escalation
        eventSubject.send(.started(escalation))

        do {
            // Create a call ID for the new voice call
            let callId = UUID().uuidString

            // Link the thread and call
            threadToCall[threadId] = callId
            callToThread[callId] = threadId

            // Update thread with linked call ID
            messagingQueueManager?.linkToCall(threadId, callId: callId)

            // Add transition message to thread
            if let thread = thread ?? messagingQueueManager?.getThread(threadId) {
                let message = reason.map {
                    "This conversation has been escalated to a voice call: \($0)"
                } ?? "This conversation has been escalated to a voice call"

                _ = try? await messagingQueueManager?.addMessage(
                    to: threadId,
                    content: message,
                    senderType: .system
                )
            }

            // Initiate the actual call
            if let service = callingService, let callerPubkey = thread?.thread.contact?.pubkey {
                try await service.startCall(to: callerPubkey, type: .voice)
            }

            // Update escalation
            escalation.callId = callId
            escalation.status = .completed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation

            eventSubject.send(.completed(escalation, callId: callId, threadId: nil))

            logger.info("Escalated thread \(threadId) to voice call \(callId)")
            return EscalationResult(
                success: true,
                escalationId: escalationId,
                newCallId: callId
            )
        } catch {
            escalation.status = .failed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation

            eventSubject.send(.failed(escalation, error.localizedDescription))

            logger.error("Failed to escalate to voice: \(error.localizedDescription)")
            return EscalationResult(
                success: false,
                escalationId: escalationId,
                error: error.localizedDescription
            )
        }
    }

    // MARK: - De-escalate to Messaging

    /// De-escalate a voice call back to messaging
    public func deescalateToMessaging(
        callId: String,
        reason: String? = nil
    ) async -> EscalationResult {
        let escalationId = UUID().uuidString
        let now = Int(Date().timeIntervalSince1970)

        // Check if there's an existing thread linked to this call
        var threadId = callToThread[callId]

        var escalation = EscalationRequest(
            id: escalationId,
            threadId: threadId ?? "",
            callId: callId,
            direction: .toMessaging,
            initiatedBy: .operator,
            reason: reason,
            status: .pending,
            createdAt: now
        )

        escalations[escalationId] = escalation
        eventSubject.send(.started(escalation))

        do {
            // If no existing thread, create a new one
            if threadId == nil {
                threadId = UUID().uuidString
                threadToCall[threadId!] = callId
                callToThread[callId] = threadId!
            }

            escalation.threadId = threadId!
            escalation.status = .completed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation

            // Add transition message
            let message = reason.map {
                "This conversation has been moved to messaging: \($0)"
            } ?? "This conversation has been moved to messaging"

            _ = try? await messagingQueueManager?.addMessage(
                to: threadId!,
                content: message,
                senderType: .system
            )

            eventSubject.send(.completed(escalation, callId: nil, threadId: threadId))

            logger.info("De-escalated call \(callId) to thread \(threadId!)")
            return EscalationResult(
                success: true,
                escalationId: escalationId,
                newThreadId: threadId
            )
        } catch {
            escalation.status = .failed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation

            eventSubject.send(.failed(escalation, error.localizedDescription))

            logger.error("Failed to de-escalate to messaging: \(error.localizedDescription)")
            return EscalationResult(
                success: false,
                escalationId: escalationId,
                error: error.localizedDescription
            )
        }
    }

    // MARK: - Caller-Initiated Escalation

    /// Request escalation from caller (caller wants to switch channels)
    public func requestEscalation(
        currentChannelId: String,
        direction: EscalationDirection,
        callerPubkey: String
    ) async -> EscalationRequest {
        let escalationId = UUID().uuidString
        let now = Int(Date().timeIntervalSince1970)

        let escalation = EscalationRequest(
            id: escalationId,
            threadId: direction == .toVoice ? currentChannelId : "",
            callId: direction == .toMessaging ? currentChannelId : nil,
            direction: direction,
            initiatedBy: .caller,
            status: .pending,
            createdAt: now
        )

        escalations[escalationId] = escalation
        updatePendingEscalations()

        eventSubject.send(.requested(request: escalation, callerPubkey: callerPubkey))

        logger.info("Caller requested escalation \(direction.rawValue)")
        return escalation
    }

    /// Accept a pending escalation request
    public func acceptEscalation(_ escalationId: String) async -> EscalationResult {
        guard var escalation = escalations[escalationId] else {
            return EscalationResult(
                success: false,
                escalationId: escalationId,
                error: "Escalation not found"
            )
        }

        guard escalation.status == .pending else {
            return EscalationResult(
                success: false,
                escalationId: escalationId,
                error: "Escalation is not pending"
            )
        }

        escalation.status = .accepted
        escalations[escalationId] = escalation

        eventSubject.send(.accepted(escalation))

        // Perform the actual escalation
        if escalation.direction == .toVoice {
            let callId = UUID().uuidString
            escalation.callId = callId
            threadToCall[escalation.threadId] = callId
            callToThread[callId] = escalation.threadId

            // Update thread with linked call
            messagingQueueManager?.linkToCall(escalation.threadId, callId: callId)

            escalation.status = .completed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation
            updatePendingEscalations()

            eventSubject.send(.completed(escalation, callId: callId, threadId: nil))

            return EscalationResult(
                success: true,
                escalationId: escalationId,
                newCallId: callId
            )
        } else {
            let threadId = UUID().uuidString
            escalation.threadId = threadId
            if let callId = escalation.callId {
                callToThread[callId] = threadId
                threadToCall[threadId] = callId
            }

            escalation.status = .completed
            escalation.completedAt = Int(Date().timeIntervalSince1970)
            escalations[escalationId] = escalation
            updatePendingEscalations()

            eventSubject.send(.completed(escalation, callId: nil, threadId: threadId))

            return EscalationResult(
                success: true,
                escalationId: escalationId,
                newThreadId: threadId
            )
        }
    }

    /// Decline a pending escalation request
    public func declineEscalation(_ escalationId: String, reason: String? = nil) -> Bool {
        guard var escalation = escalations[escalationId],
              escalation.status == .pending else {
            return false
        }

        escalation.status = .declined
        escalation.completedAt = Int(Date().timeIntervalSince1970)
        if let reason = reason {
            escalation.reason = reason
        }

        escalations[escalationId] = escalation
        updatePendingEscalations()

        eventSubject.send(.declined(escalation))

        logger.info("Declined escalation: \(escalationId)")
        return true
    }

    // MARK: - Linkage Queries

    /// Get call ID linked to a thread
    public func getLinkedCall(_ threadId: String) -> String? {
        threadToCall[threadId]
    }

    /// Get thread ID linked to a call
    public func getLinkedThread(_ callId: String) -> String? {
        callToThread[callId]
    }

    /// Check if thread has a linked call
    public func hasLinkedCall(_ threadId: String) -> Bool {
        threadToCall[threadId] != nil
    }

    /// Check if call has a linked thread
    public func hasLinkedThread(_ callId: String) -> Bool {
        callToThread[callId] != nil
    }

    /// Clear link between thread and call (on call end)
    public func clearLink(_ channelId: String) {
        // Try to find and clear by thread ID
        if let callId = threadToCall[channelId] {
            threadToCall.removeValue(forKey: channelId)
            callToThread.removeValue(forKey: callId)
            eventSubject.send(.linkCleared(channelId: channelId))
            logger.debug("Cleared link for thread: \(channelId)")
            return
        }

        // Try to find and clear by call ID
        if let threadId = callToThread[channelId] {
            callToThread.removeValue(forKey: channelId)
            threadToCall.removeValue(forKey: threadId)
            eventSubject.send(.linkCleared(channelId: channelId))
            logger.debug("Cleared link for call: \(channelId)")
        }
    }

    // MARK: - Escalation Queries

    /// Get escalation by ID
    public func getEscalation(_ escalationId: String) -> EscalationRequest? {
        escalations[escalationId]
    }

    /// Get pending escalation requests for a thread or call
    public func getPendingEscalations(for channelId: String) -> [EscalationRequest] {
        escalations.values.filter {
            $0.status == .pending &&
            ($0.threadId == channelId || $0.callId == channelId)
        }
    }

    /// Get escalation history for a channel
    public func getHistory(for channelId: String) -> [EscalationRequest] {
        escalations.values
            .filter { $0.threadId == channelId || $0.callId == channelId }
            .sorted { $0.createdAt > $1.createdAt }
    }

    // MARK: - Private Helpers

    private func updatePendingEscalations() {
        pendingEscalations = escalations.values
            .filter { $0.status == .pending }
            .sorted { $0.createdAt < $1.createdAt }
    }

    /// Create thread from call context
    public func createThreadFromCall(
        callId: String,
        callerPubkey: String,
        callerName: String? = nil,
        hotlineId: String
    ) async -> String? {
        guard let manager = messagingQueueManager else {
            logger.warning("MessagingQueueManager not available")
            return nil
        }

        do {
            let thread = try await manager.createThread(
                callerPubkey: callerPubkey,
                callerName: callerName,
                contactType: .buildit,
                initialMessage: "Call initiated - transitioning to messaging",
                priority: .medium
            )

            // Link the thread and call
            threadToCall[thread.threadId] = callId
            callToThread[callId] = thread.threadId

            // Update thread with linked call
            manager.linkToCall(thread.threadId, callId: callId)

            logger.info("Created thread \(thread.threadId) from call \(callId)")
            return thread.threadId
        } catch {
            logger.error("Failed to create thread from call: \(error.localizedDescription)")
            return nil
        }
    }

    /// Resume thread from previous call
    public func resumeThreadFromCall(
        callId: String,
        threadId: String
    ) {
        // Re-establish linkage
        threadToCall[threadId] = callId
        callToThread[callId] = threadId

        // Update thread with linked call
        messagingQueueManager?.linkToCall(threadId, callId: callId)

        logger.info("Resumed thread \(threadId) for call \(callId)")
    }
}

// MARK: - Errors

public enum EscalationError: LocalizedError {
    case notFound
    case notPending
    case escalationFailed(String)

    public var errorDescription: String? {
        switch self {
        case .notFound:
            return "Escalation not found"
        case .notPending:
            return "Escalation is not pending"
        case .escalationFailed(let reason):
            return "Escalation failed: \(reason)"
        }
    }
}

File: clients/web/src/modules/calling/services/channelEscalation.ts
/**
 * Channel Escalation Service
 * Handles transitions between messaging and voice channels
 */

import { EventEmitter } from 'events';
import type { MessagingThread } from './messagingQueueManager';

export interface EscalationRequest {
  id: string;
  threadId: string;
  callId?: string;
  direction: 'to-voice' | 'to-messaging';
  initiatedBy: 'operator' | 'caller' | 'system';
  reason?: string;
  status: 'pending' | 'accepted' | 'declined' | 'completed' | 'failed';
  createdAt: number;
  completedAt?: number;
}

export interface EscalationResult {
  success: boolean;
  escalationId: string;
  newCallId?: string;
  newThreadId?: string;
  error?: string;
}

export class ChannelEscalation extends EventEmitter {
  private escalations: Map<string, EscalationRequest> = new Map();
  private threadToCall: Map<string, string> = new Map(); // threadId -> callId
  private callToThread: Map<string, string> = new Map(); // callId -> threadId
  private operatorPubkey: string = '';

  constructor() {
    super();
  }

  /**
   * Initialize with operator context
   */
  initialize(operatorPubkey: string): void {
    this.operatorPubkey = operatorPubkey;
  }

  /**
   * Escalate a messaging thread to a voice call
   */
  async escalateToVoice(
    threadId: string,
    thread: MessagingThread,
    reason?: string
  ): Promise<EscalationResult> {
    const escalationId = crypto.randomUUID();
    const now = Date.now();

    const escalation: EscalationRequest = {
      id: escalationId,
      threadId,
      direction: 'to-voice',
      initiatedBy: 'operator',
      reason,
      status: 'pending',
      createdAt: now,
    };

    this.escalations.set(escalationId, escalation);
    this.emit('escalation-started', escalation);

    try {
      // Create a call ID for the new voice call
      const callId = crypto.randomUUID();

      // Link the thread and call
      this.threadToCall.set(threadId, callId);
      this.callToThread.set(callId, threadId);

      // In real implementation:
      // 1. Initiate WebRTC call to the caller
      // 2. Wait for caller to accept
      // 3. Transition the thread to voice mode

      escalation.callId = callId;
      escalation.status = 'completed';
      escalation.completedAt = Date.now();

      this.emit('escalation-completed', {
        escalation,
        thread,
        callId,
      });

      // Add system message to thread
      this.addTransitionMessage(thread, 'to-voice', reason);

      return {
        success: true,
        escalationId,
        newCallId: callId,
      };
    } catch (error) {
      escalation.status = 'failed';
      escalation.completedAt = Date.now();

      this.emit('escalation-failed', {
        escalation,
        error,
      });

      return {
        success: false,
        escalationId,
        error: error instanceof Error ? error.message : 'Escalation failed',
      };
    }
  }

  /**
   * De-escalate a voice call back to messaging
   */
  async deescalateToMessaging(
    callId: string,
    reason?: string
  ): Promise<EscalationResult> {
    const escalationId = crypto.randomUUID();
    const now = Date.now();

    // Check if there's an existing thread linked to this call
    let threadId = this.callToThread.get(callId);

    const escalation: EscalationRequest = {
      id: escalationId,
      threadId: threadId || '',
      callId,
      direction: 'to-messaging',
      initiatedBy: 'operator',
      reason,
      status: 'pending',
      createdAt: now,
    };

    this.escalations.set(escalationId, escalation);
    this.emit('escalation-started', escalation);

    try {
      // If no existing thread, create a new one
      if (!threadId) {
        threadId = crypto.randomUUID();
        this.threadToCall.set(threadId, callId);
        this.callToThread.set(callId, threadId);
      }

      escalation.threadId = threadId;
      escalation.status = 'completed';
      escalation.completedAt = Date.now();

      this.emit('escalation-completed', {
        escalation,
        callId,
        threadId,
      });

      return {
        success: true,
        escalationId,
        newThreadId: threadId,
      };
    } catch (error) {
      escalation.status = 'failed';
      escalation.completedAt = Date.now();

      this.emit('escalation-failed', {
        escalation,
        error,
      });

      return {
        success: false,
        escalationId,
        error: error instanceof Error ? error.message : 'De-escalation failed',
      };
    }
  }

  /**
   * Request escalation from caller (caller wants to switch channels)
   */
  async requestEscalation(
    currentChannelId: string,
    direction: 'to-voice' | 'to-messaging',
    callerPubkey: string
  ): Promise<EscalationRequest> {
    const escalationId = crypto.randomUUID();
    const now = Date.now();

    const escalation: EscalationRequest = {
      id: escalationId,
      threadId: direction === 'to-voice' ? currentChannelId : '',
      callId: direction === 'to-messaging' ? currentChannelId : undefined,
      direction,
      initiatedBy: 'caller',
      status: 'pending',
      createdAt: now,
    };

    this.escalations.set(escalationId, escalation);
    this.emit('escalation-requested', { escalation, callerPubkey });

    return escalation;
  }

  /**
   * Accept a pending escalation request
   */
  async acceptEscalation(escalationId: string): Promise<EscalationResult> {
    const escalation = this.escalations.get(escalationId);
    if (!escalation) {
      return {
        success: false,
        escalationId,
        error: 'Escalation not found',
      };
    }

    if (escalation.status !== 'pending') {
      return {
        success: false,
        escalationId,
        error: 'Escalation is not pending',
      };
    }

    escalation.status = 'accepted';

    // Perform the actual escalation
    if (escalation.direction === 'to-voice') {
      const callId = crypto.randomUUID();
      escalation.callId = callId;
      this.threadToCall.set(escalation.threadId, callId);
      this.callToThread.set(callId, escalation.threadId);

      escalation.status = 'completed';
      escalation.completedAt = Date.now();

      this.emit('escalation-completed', escalation);

      return {
        success: true,
        escalationId,
        newCallId: callId,
      };
    } else {
      const threadId = crypto.randomUUID();
      escalation.threadId = threadId;
      if (escalation.callId) {
        this.callToThread.set(escalation.callId, threadId);
        this.threadToCall.set(threadId, escalation.callId);
      }

      escalation.status = 'completed';
      escalation.completedAt = Date.now();

      this.emit('escalation-completed', escalation);

      return {
        success: true,
        escalationId,
        newThreadId: threadId,
      };
    }
  }

  /**
   * Decline a pending escalation request
   */
  declineEscalation(escalationId: string, reason?: string): boolean {
    const escalation = this.escalations.get(escalationId);
    if (!escalation || escalation.status !== 'pending') {
      return false;
    }

    escalation.status = 'declined';
    escalation.completedAt = Date.now();
    if (reason) {
      escalation.reason = reason;
    }

    this.emit('escalation-declined', escalation);
    return true;
  }

  /**
   * Get call ID linked to a thread
   */
  getLinkedCall(threadId: string): string | undefined {
    return this.threadToCall.get(threadId);
  }

  /**
   * Get thread ID linked to a call
   */
  getLinkedThread(callId: string): string | undefined {
    return this.callToThread.get(callId);
  }

  /**
   * Check if thread has a linked call
   */
  hasLinkedCall(threadId: string): boolean {
    return this.threadToCall.has(threadId);
  }

  /**
   * Check if call has a linked thread
   */
  hasLinkedThread(callId: string): boolean {
    return this.callToThread.has(callId);
  }

  /**
   * Get escalation by ID
   */
  getEscalation(escalationId: string): EscalationRequest | undefined {
    return this.escalations.get(escalationId);
  }

  /**
   * Get pending escalation requests for a thread or call
   */
  getPendingEscalations(channelId: string): EscalationRequest[] {
    return Array.from(this.escalations.values()).filter(
      (e) =>
        e.status === 'pending' &&
        (e.threadId === channelId || e.callId === channelId)
    );
  }

  /**
   * Add a system message about channel transition
   */
  private addTransitionMessage(
    thread: MessagingThread,
    direction: 'to-voice' | 'to-messaging',
    reason?: string
  ): void {
    const message = direction === 'to-voice'
      ? '📞 This conversation has been escalated to a voice call' +
        (reason ? `: ${reason}` : '')
      : '💬 This conversation has been moved to messaging' +
        (reason ? `: ${reason}` : '');

    thread.messages.push({
      id: crypto.randomUUID(),
      threadId: thread.threadId,
      content: message,
      senderPubkey: 'system',
      senderType: 'system',
      timestamp: Date.now(),
      read: true,
    });
  }

  /**
   * Clear link between thread and call (on call end)
   */
  clearLink(channelId: string): void {
    // Try to find and clear by thread ID
    const callId = this.threadToCall.get(channelId);
    if (callId) {
      this.threadToCall.delete(channelId);
      this.callToThread.delete(callId);
      return;
    }

    // Try to find and clear by call ID
    const threadId = this.callToThread.get(channelId);
    if (threadId) {
      this.callToThread.delete(channelId);
      this.threadToCall.delete(threadId);
    }
  }

  /**
   * Get escalation history for a channel
   */
  getHistory(channelId: string): EscalationRequest[] {
    return Array.from(this.escalations.values())
      .filter((e) => e.threadId === channelId || e.callId === channelId)
      .sort((a, b) => b.createdAt - a.createdAt);
  }
}

// Singleton instance
export const channelEscalation = new ChannelEscalation();
